commands/bootstrap

```diff
@@ -297,6 +332,331 @@ create_self_signed() {
+    LOCAL_CERT_NAME="${DEVICE_ID}-${SERVICE}"
+    LOCAL_CERT_FILE="${LOCAL_CERT_DIR}/${LOCAL_CERT_NAME}.crt"
+    LOCAL_CERT_KEY="${LOCAL_CERT_DIR}/${LOCAL_CERT_NAME}.key"
+    LOCAL_SIGNING_CERT_FILE="${LOCAL_CERT_DIR}/${SIGNING_CERT_NAME}.crt"
+
+    "${EXEC_CMD[@]}" "$TARGET" mkdir -p "$LOCAL_CERT_DIR"
+
+    if ! "${EXEC_CMD[@]}" "$TARGET" test -f "$LOCAL_CERT_KEY"; then
+        info "Creating device private key"
+        "${EXEC_CMD[@]}" "$TARGET" $SUDO openssl genrsa -out "$LOCAL_CERT_KEY" 2048
+        "${EXEC_CMD[@]}" "$TARGET" sudo chown tedge:root "$LOCAL_CERT_KEY"
+        "${EXEC_CMD[@]}" "$TARGET" sudo chmod 400 "$LOCAL_CERT_KEY"
+    else
+        info "Using existing private key"
+    fi
+
+    # Create Certificate Signing Request (CSR)
+    DEVICE_CSR=$(
+        "${EXEC_CMD[@]}" "$TARGET" $SUDO openssl req \
+            -key "$LOCAL_CERT_KEY" \
+            -new \
+            -subj "/O=thin-edge/OU=${DEVICE_ID}/CN=${SERVICE}"
+    )
+
+    # Sign the CSR
+    if [ "$SERVICE" = "child-agent" ]; then
+    CERT_EXT=$(cat << EOF
+basicConstraints=CA:FALSE
+extendedKeyUsage = clientAuth
+EOF
+    )
+    else
+    CERT_EXT=$(cat << EOF
+authorityKeyIdentifier=keyid
+basicConstraints=CA:FALSE
+keyUsage = digitalSignature, keyAgreement
+extendedKeyUsage = serverAuth, clientAuth
+subjectAltName = DNS:${TARGET_DNS},IP:${TARGET_IP},DNS:localhost,IP:127.0.0.1
+EOF
+    )
+    fi
+
+    DEVICE_CERT_CHAIN="/tmp/${LOCAL_CERT_NAME}.crt"
+    openssl x509 -req \
+        -in <(echo "$DEVICE_CSR") \
+        -out "$DEVICE_CERT_CHAIN" \
+        -CA "$SIGNING_CERT_FILE" \
+        -CAkey "$SIGNING_CERT_KEY" \
+        -extfile <(echo "$CERT_EXT") \
+        -CAcreateserial \
+        -days "$LEAF_EXPIRE_DAYS"
+
+    # Build certificate chain (from leaf cert to the signing cert)
+    cat "$SIGNING_CERT_FILE" >> "$DEVICE_CERT_CHAIN"
+
+    # Copy cert to device
+    "${SCP_CMD}" "$DEVICE_CERT_CHAIN" "$TARGET" "/tmp/${LOCAL_CERT_NAME}.crt"
+    "${EXEC_CMD[@]}" "$TARGET" "$SUDO" mv "/tmp/${LOCAL_CERT_NAME}.crt" "$LOCAL_CERT_FILE"
+    "${EXEC_CMD[@]}" "$TARGET" "$SUDO" chown tedge:root "$LOCAL_CERT_FILE"
+    "${EXEC_CMD[@]}" "$TARGET" "$SUDO" chmod 644 "$LOCAL_CERT_FILE"
+
+    case "$SERVICE" in
+        mqtt-broker)
+            configure_mtls_mqtt_broker "$TARGET_IP" "$TARGET_DNS" "$LOCAL_SIGNING_CERT_FILE" "$LOCAL_CERT_FILE" "$LOCAL_CERT_KEY"
+            ;;
+        main-agent)
+            configure_mtls_mqtt_client "$TARGET_IP" "$TARGET_DNS" "$LOCAL_SIGNING_CERT_FILE" "$LOCAL_CERT_FILE" "$LOCAL_CERT_KEY"
+            configure_mtls_main_agent "$TARGET_IP" "$TARGET_DNS" "$LOCAL_CERT_FILE" "$LOCAL_CERT_KEY"
+            ;;
+        c8y-mapper)
+            configure_mtls_c8y_mapper "$TARGET_IP" "$TARGET_DNS" "$LOCAL_CERT_FILE" "$LOCAL_CERT_KEY"
+            ;;
+        child-agent)
+	    # GET IP and DNS of the main device
+	    MAIN_DEVICE_DNS=$("${EXEC_CMD[@]}" "$MAIN_DEVICE" hostname 2>/dev/null || true)
```

Comment:
Using the `hostname` unfortunately does not work all the time to get a name which can be resolved in some networks, for example in my local network the hostname is only resolved with the `.local` suffix, e.g. `rpi5-abcdef01.local` and not `rpi5-abcdef01`...so we should probably add some check to see if the detected dns name is actually resolvable in the network (testing both the hostname and the hostname with .local suffix)

Reply:
A TODO has been added to the code once resolving .local address is supported by thin-edge.io, see ticket https://github.com/thin-edge/thin-edge.io/issues/2803